### DISCUSSION TITLE
Feature request: Allow for import of external types.

### DISCUSSION THREAD

**Discussion by nonetallt** [spatie/laravel-typescript-transformer#25] [Q&A]
Some background for example use scenario:  I'm currently using https://github.com/fumeapp/modeltyper to generate types for my models and I would love to use those types with types generated by this package.

Currently I have a custom transformer and a writer to help me transform my broadcasted events into typescript types (the transformer takes all broadcast events and converts all properties that are models into their short name variants, the writer checks the content for models that need to be imported.)

```php
<?php

namespace App\Domain\Typescript;

use Illuminate\Database\Eloquent\Model;
use Spatie\TypeScriptTransformer\Structures\TransformedType;
use Spatie\TypeScriptTransformer\Structures\TypesCollection;
use Spatie\TypeScriptTransformer\Writers\ModuleWriter;

class ExternalImportModuleWriter extends ModuleWriter
{
    public function format(TypesCollection $collection): string
    {
        $models = resolve(GetAllReflectionModels::class);
        $importedModels = [];

        foreach($collection as $type) {
            /** @var TransformedType $type */

            foreach($models as $shortName => $reflection) {
                if(strpos($type->transformed, $shortName . PHP_EOL) === false) {
                    continue;
                }

                $importedModels[$shortName] = $shortName;
            }
        }

        $importedModelsString = implode(',', $importedModels);
        $imports = "import {{$importedModelsString}} from '@/types/models'";

        return $imports . str_repeat(PHP_EOL, 2) . parent::format($collection);
    }
}
```

This results in the following output for my two event classes that have a single model property:

```ts
import { TaskScheduleInstanceTimer } from '@/types/models'

export type TaskTimerCompleted = {
  timer: TaskScheduleInstanceTimer
}
export type TaskTimerStarted = {
  timer: TaskScheduleInstanceTimer
}
```

However, the way I have currently achieved this is terribly clumsy. There are no properties on the transformed type that would allow me to iterate over its properties so I have to try checking the imports from the formatted typescript string which is far from ideal.

**Possible Fixes:**

Preserve an iterable list of nested types in `TransformedType` (to allow for easier access by the writer to underlying types)

and/or

 Allow for custom implementations for the [replace missing symbols step](https://spatie.be/docs/typescript-transformer/v2/under-the-hood#content-step-3-replacing-missing-symbols) of the application. This could be used to ignore replacement of missing symbols into `any` when there's a type that can be imported by the writer.

**Comment by rubenvanassche** [comment]
We have the `TypeScriptTransformableAttribute` interface which can be used to output custom types using attributes. Wouldn't that solve your problem?
